optimum/onnxruntime/io_binding/io_binding_helper.py

```diff
@@ -0,0 +1,159 @@
+
+    @staticmethod
+    def ort_type_to_torch_type(ort_type: str):
+        ort_type_to_torch_type_map = {
+            "tensor(int64)": torch.int64,
+            "tensor(int32)": torch.int32,
+            "tensor(int8)": torch.int8,
+            "tensor(float)": torch.float32,
+            "tensor(float16)": torch.float16,
+            "tensor(bool)": torch.bool,
+        }
+        if ort_type not in ort_type_to_torch_type_map:
+            raise ValueError(f"{ort_type} not found in map")
+        return ort_type_to_torch_type_map[ort_type]
+
+
+class IOBindingHelper:
+    def __init__(self, model: ort.InferenceSession, config, device, **kwargs):
+        self.model = model
+        self.config = config
+        self.device = device
+        # create {name:idx} dict for model outputs
+        self.model_inputs = {output_key.name: idx for idx, output_key in enumerate(model.get_inputs())}
+        self.model_outputs = {output_key.name: idx for idx, output_key in enumerate(model.get_outputs())}
+        self.model_input_names = list(self.model_inputs.keys())
+        self.model_output_names = list(self.model_outputs.keys())
+
+    def prepare_io_binding(self, **kwargs):
+        """Returnas IO binding object for a session."""
+
+        name_to_np_type = TypeHelper.get_io_numpy_type_map(self.model)
+
+        # Bind inputs and outputs to onnxruntime session
+        io_binding = self.model.io_binding()
+
+        # Bind inputs
+        for input_name in self.model_input_names:
+            onnx_input = kwargs.pop(input_name)
+
+            assert onnx_input.is_contiguous()
+            io_binding.bind_input(
+                input_name,
+                onnx_input.device.type,
+                self.device.index,
+                name_to_np_type[input_name],
+                list(onnx_input.size()),
+                onnx_input.data_ptr(),
+            )
+
+        # Bind outputs
+        for name in self.model_output_names:
+            io_binding.bind_output(name, self.device.type, device_id=self.device.index)
+
+        return io_binding
+
+    @staticmethod
+    def to_pytorch(ort_value: OrtValue) -> torch.Tensor:
+        env = {pkg.key for pkg in pkg_resources.working_set}
+        if "onnxruntime-training" in env:
+            return IOBindingHelper.to_pytorch_via_dlpack(ort_value)
+        else:
+            return IOBindingHelper.to_pytorch_via_cupy(ort_value)
+            # try:
+            #     return IOBindingHelper.to_pytorch_via_cupy(ort_value)
+            # except Exception as e:
+            #     logging.error(traceback.format_exc())
+            #     logging.info("Unable to access output memory in CUDA, will offload to CPU")
+            #     return IOBindingHelper.to_pytorch_via_np(ort_value)
+
+    @staticmethod
+    def to_pytorch_via_np(ort_value: OrtValue) -> torch.Tensor:
+        ort_device = ort_value.device_name().lower()
+        return torch.tensor(ort_value.numpy()).to(ort_device)
```

Comment:
When ort offloads the data to CPU with numpy, I need to turn numpy to torch tensor and send it back to GPU. Otherwise, during generation, the model is still on GPU while logit is in CPU, and it will fail.